### PR TITLE
Lazy Tensor to `torch` MLIR

### PR DESCRIPTION
**Pull request by ramiro050**
This PR implements a way of compiling a lazy computation captured by the [Lazy Tensor Core](https://github.com/pytorch/pytorch/tree/lazy_tensor_staging/lazy_tensor_core) (LTC) of Pytorch into the `torch` MLIR dialect. The following is a high level description of how this compilation is done:

1. Use the built-in TorchScript plugin of LTC to convert the lazy computation captured by LTC into a `torch.jit.ScriptFunction`
2. Use the `ModuleBuilder.import_function` to convert the `ScriptFunction` into an `MLIR` module
3. Use the the new pass for `ScriptFunction`s in the `torch_mlir.passmanager` to compile the MLIR module all the way down

## Request for feedback

- I am currently modifying the pass pipelines of `torch-mlir` (see [commit](https://github.com/ramiro050/mlir-npcomp/commit/cb1fac76c969cfd6adaefca6382f05db5b84a43d)). The reason this is needed is because the `GlobalizedModule` pass expects a `torch.nn.Module` to be part of the MLIR, which is always true for `torch.jit.ScriptModule` objects, but not for `torch.jit.ScriptFunction` objects. Thus, I decided to turn the existing TorchScript pipeline into two pipelines, one for `ScriptFunction`s and one for `ScriptModule`s. The way I chose to change things works for compiling `torch.jit.ScriptFunction`s, but I am not sure if it is completely correct or if there is a better way of doing it. I would love any feedback on this change.

## Under construction

- `utils/annotator.py`: This file is currently a work in progress. The next step is to change the `torchfx` code so that it uses a similar interface to the one `lazytensor` is using. In particular, the `Builder` used to turn `torch.fx.ModuleGraph` into MLIR will in the future take an extra argument for an `Annotation` object, and the `annotate_forward_args` function in `annotate.py` will be removed
- I currently have to convert the lazy computation to a string then parse it back in to get a TorchScript IR graph. This should be easily fixable by modifying Pytorch, which I plan on doing in the future

### CODE REVIEW

File: examples/lazytensor_tanh.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+print("MLIR")
+mlir_module.dump()
+
+# Compile the torch MLIR using npcomp and execute the compiled program
+with mlir_module.context:
+    pm = PassManager.parse('torchscript-function-to-linalg-on-tensors-backend-pipeline')
+pm.run(mlir_module)
+
+print("BEFORE NPCOMP BACKEND PIPELINE")
```
Comment (silvasean) [inline examples/lazytensor_tanh.py:68]: nit: remove the word "npcomp" (we are avoiding using that going forward)

File: examples/torchfx/builder.py [hunk 2]
```diff
@@ -10,8 +10,9 @@
 to convert the output from the tracer into MLIR using the `torch`
 dialect.
 
-This file is licensed under a pytorch-style license
-See frontends/pytorch/LICENSE for license information.
+Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
```
Comment (silvasean) [inline examples/torchfx/builder.py:13]: nit: would recommend keeping the license in `#` style comments at the top, followed by the docstring.

(thanks for the docstrings btw :) )

File: examples/lazytensor_tanh.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+ltc._LAZYC._ltc_init_ts_backend()
+device = 'lazy'
+dtype = torch.float32
+shape = (2, 3)
+
+x = torch.randn(shape, device=device, dtype=dtype)
+y = torch.randn(shape, device=device, dtype=dtype)
+
+# Capture lazy computation and convert to TorchScript IR
+graph_str = ltc._LAZYC._get_ltc_tensors_backend([x.tanh() * y.tanh()])
```
Comment (silvasean) [inline examples/lazytensor_tanh.py:42]: what determines the order of arguments to the TorchScript function that gets created? It seems like nothing here knows whether we want `x` or `y` to be the first argument.

**Review by silvasean — APPROVED** [review]
great work!

**Comment by shabalind** [comment]
It's great to see LazyTensor+MLIR being connected together! 👏

File: examples/lazytensor_tanh.py (hunk 3 above)
Comment (ramiro050) [reply examples/lazytensor_tanh.py]: I am not completely sure. From experimentation, the evaluation order does not seem to be from left to right all the time. For example, `x.tanh() * y` captures `y` as the first argument. I will look into ways of controlling the order

**Comment by ramiro050** [comment]
## PR Update

- [x] Update `README`
- [x] Change license style to `#` comments
- [x] Change description of `torchscript-function` pipeline
- [x] Remove mentions of `npcomp`
- [x] Added more documentation
- [x] Fixed a few bugs
- [x] Minor style fixes

**Comment by ramiro050** [comment]
Forgot to update e2e tests to use the new backend name: `torchscript-module-to...`. I'll fix it soon

**Comment by ramiro050** [comment]
Done!

**Pull request merged**